shedlock-provider-hazelcast/src/main/java/net/javacrumbs/shedlock/provider/hazelcast/HazelcastLockProvider.java

```diff
@@ -0,0 +1,191 @@
+        final String lockName = lockConfiguration.getName();
+        final IMap<String, HazelcastLock> store = getStore();
+        try {
+            // lock the map key entry
+            store.lock(lockName);
+            // just one thread at a time, in the cluster, can run this code
+            // each thread waits until the lock to be unlock
+            if (tryLock(lockConfiguration, now)) {
+                return Optional.of(() -> unlock(lockName));
+            }
+        } finally {
+            // released the map lock for the others threads
+            store.unlock(lockName);
+        }
+        return Optional.empty();
+    }
+
+    private boolean tryLock(final LockConfiguration lockConfiguration, final Instant now) {
+        final String lockName = lockConfiguration.getName();
+        final HazelcastLock lock = getLock(lockName);
+        if (isUnlocked(lock)) {
+            log.debug("lock - lock obtained, it wasn't locked : conf={}", lockConfiguration);
+            addNewLock(lockConfiguration);
+            return true;
+        } else if (isExpired(lock, now)) {
+            log.debug("lock - lock obtained, it was locked but expired : oldLock={};  conf={}", lock, lockConfiguration);
+            removeLock(lockName);
+            addNewLock(lockConfiguration);
+            return true;
+        } else if (isLockedByUnvailableMemberOfCluster(lock)) {
+            log.debug("lock - lock obtained, it was locked by an available member of cluster :  oldLock={};  conf={}", lock, lockConfiguration);
+            removeLock(lockName);
+            addNewLock(lockConfiguration);
+            return true;
+        } else {
+            log.debug("lock - already locked : currentLock={};  conf={}", lock, lockConfiguration);
+            return false;
+        }
+    }
+
+    private IMap<String, HazelcastLock> getStore() {
+        return hazelcastInstance.getMap(lockStoreKey);
+    }
+
+    protected HazelcastLock getLock(final String lockName) {
+        final IMap<String, HazelcastLock> store = getStore();
+        return store.get(lockName);
+    }
+
+    private void removeLock(final String lockName) {
+        final IMap<String, HazelcastLock> store = getStore();
+        store.delete(lockName);
+        log.debug("lock store - lock deleted : {}", lockName);
+    }
+
+    private void addNewLock(final LockConfiguration lockConfiguration) {
+        final String localMemberUuid = getLocalMemberUuid();
+        final HazelcastLock lock = HazelcastLock.fromLockConfiguration(lockConfiguration, localMemberUuid);
+        final String lockName = lockConfiguration.getName();
+        final IMap<String, HazelcastLock> store = getStore();
+        store.put(lockName, lock);
+        log.debug("lock store - new lock added : {}", lock);
+    }
+
+    private String getLocalMemberUuid() {
+        return hazelcastInstance.getCluster().getLocalMember().getUuid();
+    }
+
+    protected boolean isUnlocked(final HazelcastLock lock) {
```

Comment:
I would prefer to keep this methods private at least until the code is stabilized.